### ISSUE TITLE
The yaml filter result is wrapped in json

### ISSUE THREAD

**Issue by nsheff** [pepkit/pephub#30]
http://pephub.databio.org/pep/geo/gse74180/convert?filter=yaml

Output of the yaml filter is an escaped string representation of a yaml inside of json:

![image](https://user-images.githubusercontent.com/5271009/168892413-c95dad25-bee4-4a7f-a939-8e8a16bfef61.png)

Instead, it should just return the yaml file.

**Comment by nsheff** [comment]
sample issue for the csv filter:

![image](https://user-images.githubusercontent.com/5271009/168892700-17104865-4111-4e56-b855-0e6a58e8d3c2.png)

**Comment by nleroy917** [comment]
That makes sense to me, we can change it to just give you a file.

**Comment by nleroy917** [comment]
How do we handle file naming? should we request a `paths` argument? Or maybe a `filenames` argument? I'm wondering if this needs to be a `POST` endpoint.

**Comment by nsheff** [comment]
I think the filename can just be some default value. The user can specify something through whatever interface they are using to download it already, I don't see why we need to duplicate that. We're just providing a default.

For default, you can use the project identifier, or something generic.

**Comment by nleroy917** [comment]
**Infrastructure Meeting 05/23/2022:**
- `POST` request doesn't make sense as much
- Make a default name?
- Filename can correspond to the key of the object
- Maybe `eido` could take as a variable an extension for each object returned
- Return just the text -> do nothing

**Comment by nsheff** [comment]
This "CSV" output is still returning it wrapped in a JSON.

http://pephub.databio.org/pep/nfcore/demo_rna_pep/convert?filter=csv

Is this the intended behavior? I was expecting it to give it to me in csv format.

**Comment by nleroy917** [comment]
This is because the filters return `str` objects, and I am just dumping that to JSON, I believe. I need to look more into it.

**Comment by nleroy917** [comment]
I'm now just returning a zip file which seems to work well and is more intuitive.

**Comment by nsheff** [comment]
why wouldn't you just return yaml in plain text?

**Comment by nleroy917** [comment]
I think because some filters (or future filters) might return mixed results in unknown formats, so the zip file was a way to nicely capture them all. The `str` representations of the results was yielding odd results. I can try some more.

**Comment by nsheff** [comment]
Well, I can say from my perspective that 99.9% of API endpoints I've interacted with return plain text. This makes it easy to consume the return value.

The only times I've thought of things returning something compressed like that is when it's a large file, to save transfer. But in this case we're not talking about a large file.

Maybe there can be a `?format` query parameter, with `=zip` returning a zip file no matter what? I think the zip file may make sense for filters that produce multiple files, but it does make it less convenient for the filters that produce only a single file if you make it universally zip only... `?format=plain` would be the default, but it would not be an option for filters that return multiple files?

**Comment by nleroy917** [comment]
The fastapi `PlainTextResponse` works wonderfully for filters that have a single return value/key. But when there are two, you cannot encode the `dict` as plain text. I guess I could just append the results to each other and separate them with a newline? I agree that returning a zip file seems non-intuitive for an API

**Comment by nsheff** [comment]
Well, that could work. I still think I'd prefer my suggestion above, though.

**Comment by nleroy917** [comment]
Oh, I didn't understand at first. I see now. I like that idea. If `?format=plain` and there are multiple files, how does one handle that case server-side? Append strings together?

**Comment by nleroy917** [comment]
How about this?
```python
@router.get("/convert")
async def convert_pep(
    proj: peppy.Project = Depends(validate_pep), 
    filter: Optional[str] = "basic",
    format: Optional[str] = "plain"
):
    """
    Convert a PEP to a specific format, f. For a list of available formats/filters,
    see /eido/filters.

    See, http://eido.databio.org/en/latest/filters/#convert-a-pep-into-an-alternative-format-with-a-filter
    for more information.
    """
    # default to basic
    if filter is None:
        filter = "basic"  # default to basic

    # validate filter exists
    filter_list = eido.get_available_pep_filters()
    if filter not in filter_list:
        raise HTTPException(
            400, f"Unknown filter '{filter}'. Available filters: {filter_list}"
        )

    # generate result
    conv_result = eido.run_filter(proj, filter, verbose=False)

    format_list = ["plain", "zip"]
    if format not in format_list:
        raise HTTPException(
            400, f"Unknown format '{format}'. Availble formats: {format_list}"
        )
    
    if format == "plain":
        return_str = "\n".join([conv_result[k] for k in conv_result])
        resp_obj = PlainTextResponse(return_str)
    else:
        resp_obj = zip_conv_result(conv_result) # returns zip file in Response() object

    return resp_obj
```
Testing the endpoint I like this approach much better

**Comment by nsheff** [comment]
Yes I think this is getting in the right direction.

One hesitation I have is for the multi-file, `format=plain` case, it won't be possible to identify the different components. In that case, it might make more sense to go back to the JSON format, so you can have an handle to refer to each component of the file, rather than just separating them with newline, which will make them hard to parse. Or -- maybe this is a third option, `?format=json`, which is basically how you were doing it originally?

One other question: for the multi-file, `format=zip` case, the files should be included as separate files in the zip archive, right?

**Comment by nleroy917** [comment]
> One other question: for the multi-file, `format=zip` case, the files should be included as separate files in the zip archive, right?

Yes. Right now it's a separate file for each returned result. 

For the `json` option, we can do that, it just comes back sort of ugly since it's embedding a string into the `json` result.

**Comment by nsheff** [comment]
> For the `json` option, we can do that, it just comes back sort of ugly since it's embedding a string into the `json` result.

Yes, but the ugliness may be justified if it's a mutli-file output, I guess. Otherwise, I'm not sure it will be useful. you'll be sticking together a CSV and a yaml file, for example, in one big blob... nothing will be able to parse that...

**Comment by nleroy917** [comment]
Ok. So, we need to make a third `json` format option?
